solutions/cses-1139.mdx

```diff
@@ -0,0 +1,113 @@
+---
+id: cses-1139
+source: CSES
+title: Distinct Colors
+author: Timothy Gao
+---
+
+## Editorial
+
+Let us consider the Euler Tour of the tree. We flatten the tree into an array, where each node corresponds to a range in this array. Now, we are essentially tasked to find the number of distinct values in a range $N$ times(once for each node).
+
+There are a variety of ways to approach the problem from here.
+
+<h3>Solution: BIT</h3>
+
+One such solution is with a binary indexed tree. Let us consider iterating through the Euler tour array from left to right. 
+When we consider a node at index $i$ in the Euler tour array, its subtree range will be $[j, i]$, where $j<=i$.
+Now, let's focus on a single color. Notice that if there are multiple colors $c$ to the left of a certain index $i$,
+only $c$'s most rightmost occurrence is relevant. This is because any non-rightmost occurrence of $c$
+from $i$ must include the rightmost occurrence of $c$. More formally, 
+any segment $[l, r]$ must contains $i$ if $l <= i <= r$. 
+
+With this observation, we can essentially reduce the distinct colors queries to a simple range query. As we iterate through the Euler Tour from left to right, mark the current index(in the BIT) as 1, and if the color of the node at the current index has occurred before, make it 0 in the BIT. We can find the solution to every node by doing a sum query
+ as we iterate.
+
+**Time Complexity:** $\mathcal O(N\log N)$
+
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+ 
+using namespace std;
+#define pb push_back
+
+struct BIT {
+		vector<int> bit;
+		int n;
+		BIT(int n) {
+				this->n = n + 1;
+				bit.resize(n+1);
+		}
+		int sum(int r) {
+				r++;
+				int ret = 0;
+				while (r > 0) {
+						ret += bit[r];
+						r -= r & -r;
+				}
+				return ret;
+		}
+		void update(int idx, int v) {
+				idx++;
+				while (idx < n) {
+						bit[idx] += v;
+						idx += idx & -idx;
+				}
+		}
+};
+
+const int MAXN = 2e5+1;
+
+// Euler Tour, color of each node, answer for each node, left endpoint of each node(as explained in editorial)
+int tour[MAXN], color[MAXN], answer[MAXN], lend[MAXN];
+
+vector<int> adj[MAXN];
+int idx = 0;
+
+void dfs(int u, int par = 0){
```

Comment:
brace style is not consisten w/ bit template

Reply:
wait wut does this mean?

Reply:
fixed